{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5a582e1a-e6da-419e-82d5-378642f5b4f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "143f9258-eb42-4e24-93f5-46684de439d2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21043/3381271953.py:1: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mouse_feat = pd.read_csv(\"GCF_000001635.27_GRCm39_feature_table.txt\", sep = \"\\t\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># feature</th>\n",
       "      <th>class</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>+</td>\n",
       "      <td>predicted gene, 26206</td>\n",
       "      <td>115487594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ncRNA</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>+</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "      <td>115487594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>3269956</td>\n",
       "      <td>3741733</td>\n",
       "      <td>-</td>\n",
       "      <td>X-linked Kx blood group related 4</td>\n",
       "      <td>497097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>3269956</td>\n",
       "      <td>3741733</td>\n",
       "      <td>-</td>\n",
       "      <td>X-linked Kx blood group related 4, transcript ...</td>\n",
       "      <td>497097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>3284705</td>\n",
       "      <td>3741721</td>\n",
       "      <td>-</td>\n",
       "      <td>X-linked Kx blood group related 4</td>\n",
       "      <td>497097</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  # feature           class genomic_accession    start      end strand  \\\n",
       "0      gene           snRNA       NC_000067.7  3172239  3172348      +   \n",
       "1     ncRNA           snRNA       NC_000067.7  3172239  3172348      +   \n",
       "2      gene  protein_coding       NC_000067.7  3269956  3741733      -   \n",
       "3      mRNA             NaN       NC_000067.7  3269956  3741733      -   \n",
       "4      mRNA             NaN       NC_000067.7  3284705  3741721      -   \n",
       "\n",
       "                                                name     GeneID  \n",
       "0                              predicted gene, 26206  115487594  \n",
       "1                                U6 spliceosomal RNA  115487594  \n",
       "2                  X-linked Kx blood group related 4     497097  \n",
       "3  X-linked Kx blood group related 4, transcript ...     497097  \n",
       "4                  X-linked Kx blood group related 4     497097  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_feat = pd.read_csv(\"GCF_000001635.27_GRCm39_feature_table.txt\", sep = \"\\t\")\n",
    "mouse_feat.drop([\"assembly_unit\", \"seq_type\", \"chromosome\", \"assembly\", \"product_accession\", \"non-redundant_refseq\", \"related_accession\", \"symbol\", \"locus_tag\", \"product_length\", \"attributes\", \"feature_interval_length\"], axis = 1, inplace = True)\n",
    "mouse_feat[\"start\"] = mouse_feat[\"start\"].astype(int)\n",
    "mouse_feat[\"end\"] = mouse_feat[\"end\"].astype(int)\n",
    "mouse_feat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c54b7de-5cce-4153-96a7-af0bc6f426ff",
   "metadata": {},
   "source": [
    "## Not in BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2d205b20-0c1d-4dbf-92c7-c9b82ddf3615",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>12037138</td>\n",
       "      <td>12037187</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>63010864</td>\n",
       "      <td>63010913</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>63011096</td>\n",
       "      <td>63011145</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>63011328</td>\n",
       "      <td>63011377</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.7</td>\n",
       "      <td>63259775</td>\n",
       "      <td>63259824</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>NC_000086.8</td>\n",
       "      <td>131860700</td>\n",
       "      <td>131860749</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>NC_000086.8</td>\n",
       "      <td>157989095</td>\n",
       "      <td>157989144</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>NC_000086.8</td>\n",
       "      <td>164792349</td>\n",
       "      <td>164792398</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>NC_000086.8</td>\n",
       "      <td>164792436</td>\n",
       "      <td>164792485</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>NC_000087.8</td>\n",
       "      <td>1902104</td>\n",
       "      <td>1902153</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    genomic_accession      start        end strand\n",
       "0         NC_000067.7   12037138   12037187      -\n",
       "1         NC_000067.7   63010864   63010913      -\n",
       "2         NC_000067.7   63011096   63011145      -\n",
       "3         NC_000067.7   63011328   63011377      -\n",
       "4         NC_000067.7   63259775   63259824      +\n",
       "..                ...        ...        ...    ...\n",
       "143       NC_000086.8  131860700  131860749      -\n",
       "144       NC_000086.8  157989095  157989144      -\n",
       "145       NC_000086.8  164792349  164792398      -\n",
       "146       NC_000086.8  164792436  164792485      -\n",
       "147       NC_000087.8    1902104    1902153      -\n",
       "\n",
       "[148 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matches = pd.read_csv(\"blast_masked/mouse_vs_fly_e1e-07_k12.fn.gff\", sep = \"\\t\", header = None)\n",
    "matches.drop([1, 2, 5, 7], inplace = True, axis = 1)\n",
    "\n",
    "if (len(matches[8].iloc[0].split(';')) == 4):\n",
    "    matches[['qname','seq2Range','cigar', 'mutations']] = matches[8].str.split(';',expand=True)\n",
    "if (len(matches[8].iloc[0].split(';')) == 5):\n",
    "    matches[['qname','seq2Range', 'eValue', 'cigar', 'mutations']] = matches[8].str.split(';',expand=True)\n",
    "    matches.drop(['eValue'], inplace = True, axis = 1)\n",
    "\n",
    "matches.drop([\"cigar\", \"mutations\"], inplace = True, axis = 1)\n",
    "matches.drop([\"seq2Range\", \"qname\", 8], inplace = True, axis = 1)\n",
    "matches.columns = [\"genomic_accession\", \"start\", \"end\", \"strand\"]\n",
    "    \n",
    "matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6a9e2af9-410d-4f5e-b5a4-daacc940e43d",
   "metadata": {},
   "outputs": [],
   "source": [
    "matches = matches.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ee3ec3b3-6700-4f71-9af4-d3844c2550a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Eps-match\tNC_000067.7\t12037138\t12037187\n",
      "\n",
      "    # feature           class genomic_accession     start       end strand  \\\n",
      "522      gene  protein_coding       NC_000067.7  11484121  12046126      +   \n",
      "523      mRNA             NaN       NC_000067.7  11484121  12046126      +   \n",
      "524      mRNA             NaN       NC_000067.7  11484121  12046126      +   \n",
      "525      mRNA             NaN       NC_000067.7  11484121  12046126      +   \n",
      "526      mRNA             NaN       NC_000067.7  11484121  12046126      +   \n",
      "527  misc_RNA             NaN       NC_000067.7  11484121  12045302      +   \n",
      "528      mRNA             NaN       NC_000067.7  11484202  12046126      +   \n",
      "529      mRNA             NaN       NC_000067.7  11484202  12046126      +   \n",
      "530      mRNA             NaN       NC_000067.7  11484240  12046126      +   \n",
      "532      mRNA             NaN       NC_000067.7  11484329  12046126      +   \n",
      "533      mRNA             NaN       NC_000067.7  11484329  12046126      +   \n",
      "534      mRNA             NaN       NC_000067.7  11484329  12046126      +   \n",
      "535      mRNA             NaN       NC_000067.7  11484329  12046126      +   \n",
      "536      mRNA             NaN       NC_000067.7  11484329  12046126      +   \n",
      "543       CDS    with_protein       NC_000067.7  11484793  12045190      +   \n",
      "544       CDS    with_protein       NC_000067.7  11484793  12045190      +   \n",
      "545       CDS    with_protein       NC_000067.7  11484793  12045190      +   \n",
      "546       CDS    with_protein       NC_000067.7  11484793  12045190      +   \n",
      "547       CDS    with_protein       NC_000067.7  11484793  12045190      +   \n",
      "548       CDS    with_protein       NC_000067.7  11484793  12045190      +   \n",
      "549       CDS    with_protein       NC_000067.7  11484793  12045190      +   \n",
      "554       CDS    with_protein       NC_000067.7  11588852  12045190      +   \n",
      "555       CDS    with_protein       NC_000067.7  11588852  12045190      +   \n",
      "557       CDS    with_protein       NC_000067.7  11658668  12045190      +   \n",
      "558       CDS    with_protein       NC_000067.7  11658668  12045190      +   \n",
      "559       CDS    with_protein       NC_000067.7  11658668  12045190      +   \n",
      "\n",
      "                                                  name  GeneID  \n",
      "522                         RIKEN cDNA A830018L16 gene  320492  \n",
      "523  RIKEN cDNA A830018L16 gene, transcript variant X7  320492  \n",
      "524  RIKEN cDNA A830018L16 gene, transcript variant X1  320492  \n",
      "525  RIKEN cDNA A830018L16 gene, transcript variant X3  320492  \n",
      "526  RIKEN cDNA A830018L16 gene, transcript variant X2  320492  \n",
      "527  RIKEN cDNA A830018L16 gene, transcript variant X4  320492  \n",
      "528  RIKEN cDNA A830018L16 gene, transcript variant X9  320492  \n",
      "529  RIKEN cDNA A830018L16 gene, transcript variant X8  320492  \n",
      "530  RIKEN cDNA A830018L16 gene, transcript variant X5  320492  \n",
      "532   RIKEN cDNA A830018L16 gene, transcript variant 1  320492  \n",
      "533   RIKEN cDNA A830018L16 gene, transcript variant 6  320492  \n",
      "534   RIKEN cDNA A830018L16 gene, transcript variant 2  320492  \n",
      "535   RIKEN cDNA A830018L16 gene, transcript variant 5  320492  \n",
      "536   RIKEN cDNA A830018L16 gene, transcript variant 7  320492  \n",
      "543  uncharacterized protein C8orf34 homolog isoform 1  320492  \n",
      "544  uncharacterized protein C8orf34 homolog isoform 6  320492  \n",
      "545  uncharacterized protein C8orf34 homolog isoform 2  320492  \n",
      "546  uncharacterized protein C8orf34 homolog isoform 5  320492  \n",
      "547  uncharacterized protein C8orf34 homolog isofor...  320492  \n",
      "548  uncharacterized protein C8orf34 homolog isofor...  320492  \n",
      "549  uncharacterized protein C8orf34 homolog isofor...  320492  \n",
      "554  uncharacterized protein C8orf34 homolog isoform 7  320492  \n",
      "555  uncharacterized protein C8orf34 homolog isofor...  320492  \n",
      "557  uncharacterized protein C8orf34 homolog isofor...  320492  \n",
      "558  uncharacterized protein C8orf34 homolog isofor...  320492  \n",
      "559  uncharacterized protein C8orf34 homolog isofor...  320492  \n",
      "\n",
      "Eps-match\tNC_000067.7\t63259775\t63259824\n",
      "\n",
      "     # feature           class genomic_accession     start       end strand  \\\n",
      "4817      gene  protein_coding       NC_000067.7  63239517  63353735      +   \n",
      "4818      mRNA             NaN       NC_000067.7  63239517  63353734      +   \n",
      "4819      mRNA             NaN       NC_000067.7  63239517  63353734      +   \n",
      "\n",
      "                                                   name  GeneID  \n",
      "4817                 zinc finger, DBF-type containing 2   73884  \n",
      "4818  zinc finger, DBF-type containing 2, transcript...   73884  \n",
      "4819  zinc finger, DBF-type containing 2, transcript...   73884  \n",
      "\n",
      "2 overlap genomic features\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for index, row in matches.iterrows():\n",
    "    dname = row[\"genomic_accession\"]\n",
    "    start = row[\"start\"]\n",
    "    end = row[\"end\"]\n",
    "    overlapping = mouse_feat[(mouse_feat[\"genomic_accession\"] == row[\"genomic_accession\"]) & (start >= mouse_feat[\"start\"]) & (start <= mouse_feat[\"end\"])] \n",
    "    if (len(overlapping) > 0):\n",
    "        i += 1\n",
    "        print(\"Eps-match\\t\" + str(dname) + \"\\t\" + str(start) + \"\\t\" + str(end) + \"\\n\")\n",
    "        print(overlapping)\n",
    "        print()\n",
    "\n",
    "print(str(i) + \" overlap genomic features\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
